File: 24_game.rb
require 'bigdecimal'

def judge_point24(nums)
  raise "Wrong num of nums" if nums.size != 4

  ops = %w{+ - * /}
  nums.permutation(4).each do |arr|
    ops.each do |op1|
      ops.each do |op2|
        ops.each do |op3|
          r = []

          begin
            r << eval("(BigDecimal(#{arr[0]})#{op1}BigDecimal(#{arr[1]})) #{op2}BigDecimal(#{arr[2]})#{op3}BigDecimal(#{arr[3]})")
          rescue
          end

          begin
            r << eval("BigDecimal(#{arr[0]})#{op1} (BigDecimal(#{arr[1]})#{op2}BigDecimal(#{arr[2]})) #{op3}BigDecimal(#{arr[3]})")
          rescue
          end

          begin
            r << eval("BigDecimal(#{arr[0]})#{op1}BigDecimal(#{arr[1]})#{op2} (BigDecimal(#{arr[2]})#{op3}BigDecimal(#{arr[3]}))")
          rescue
          end

          begin
            r << eval("(BigDecimal(#{arr[0]})#{op1}BigDecimal(#{arr[1]})) #{op2} (BigDecimal(#{arr[2]})#{op3}BigDecimal(#{arr[3]}))")
          rescue
          end

          begin
            r << eval("(BigDecimal(#{arr[0]})#{op1}BigDecimal(#{arr[1]})#{op2}BigDecimal(#{arr[2]})) #{op3}BigDecimal(#{arr[3]})")
          rescue
          end

          begin
            r << eval("BigDecimal(#{arr[0]})#{op1}(BigDecimal(#{arr[1]})#{op2}BigDecimal(#{arr[2]})#{op3}BigDecimal(#{arr[3]}))")
          rescue
          end

          return true if r.select {|i| (i > 23.9 and i < 24.0) or i == 24 }.size > 0
          # return true if r.include? 24
        end
      end
    end
  end

  false
end

pp true == judge_point24([4,1,8,7])
pp true == judge_point24([1,9,1,2])
pp false == judge_point24([1,2,1,2])
pp true == judge_point24([3,9,7,7])
pp true == judge_point24([8,1,6,6])
pp true == judge_point24([1,1,9,6])
pp false == judge_point24([3,4,6,7])
pp true == judge_point24([3,3,8,8])
pp true == judge_point24([3,3,7,7])

File: 24_game.go
package main

import "fmt"

func judgePoint24(nums []float32) bool {
	return dfs(nums, []float32{}, []string{}, len(nums))
}

func dfs(remain []float32, current []float32, ops []string, size int) bool {
	if len(current) == size {
		result := calculate(current, ops)
		// fmt.Println(current, ops, result)
		return result == 24
	}

	newRemain := []float32{}
	newCurrent := []float32{}
	newOps := []string{}

	for i, num := range remain {
		newCurrent = current
		newCurrent = append(newCurrent, num)

		newRemain = []float32{}
		newRemain = append(newRemain, remain[:i]...)
		newRemain = append(newRemain, remain[i+1:]...)

		for _, op := range []string{"+", "-", "*", "/"} {
			newOps = ops
			if len(newRemain) != 0 {
				newOps = append(newOps, op)
			}
			if dfs(newRemain, newCurrent, newOps, size) {
				return true
			}
		}
	}
	return false
}

var funcs = map[string]func(float32, float32) float32{
	"+": func(a, b float32) float32 { return a + b },
	"-": func(a, b float32) float32 { return a - b },
	"*": func(a, b float32) float32 { return a * b },
	"/": func(a, b float32) float32 { return a / b },
}

func calculate(nums []float32, ops []string) float32 {
	if len(nums) == 0 {
		return 0
	} else if len(nums) == 1 {
		return float32(nums[0])
	} else if len(nums) == 2 {
		return funcs[ops[0]](float32(nums[0]), float32(nums[1]))
	}

	result := calculate(nums[:2], ops[:2])
	return calculate(append([]float32{result}, nums[2:]...), ops[1:])
}

func main() {
	testCases := [][]float32{
		[]float32{1, 2, 3, 4},
		[]float32{8, 4, 6, 6},
		[]float32{9, 7, 5, 2},
	}

	for _, arr := range testCases {
		fmt.Println(judgePoint24(arr))
	}
}
